### PR TITLE
refactor(scrub): replace umoci logic in scrub implementation

### PR DESCRIPTION
**Pull request by Andreea-Lupu**
- implement scrub also for S3 storage by replacing umoci
- change scrub implementation for ImageIndex
- take the `Subject` into consideration when running scrub
- remove test code relying on the umoci library. Since we started
relying on images in test/data, and we create our own images using
go code we can obtain digests by other means. (cherry picked from commit https://github.com/andaaron/zot/commit/489d4e2d23c1b4e48799283f8281024bbef6123f)

**What type of PR is this?**

**Which issue does this PR fix**:
#1685 #734 

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/storage/scrub.go [hunk 1]
```diff
@@ -92,84 +88,171 @@ func CheckRepo(ctx context.Context, imageName string, imgStore storageTypes.Imag
...
+		results = append(results, imageResult)
+	}
+
+	return results, nil
+}
+
+func getManifestResult(
+	ctx context.Context, manifest ispec.Descriptor, imgStore storageTypes.ImageStore, imageName, tag string,
+	listOfManifests map[godigest.Digest]ScrubImageResult,
+) ScrubImageResult {
```
Comment (andaaron) [inline pkg/storage/scrub.go:131]: Why do we need to return the results, if we add them to the map anyway?

File: pkg/storage/scrub.go [hunk 2]
```diff
@@ -92,84 +88,171 @@ func CheckRepo(ctx context.Context, imageName string, imgStore storageTypes.Imag
...
-		results = append(results, imageResult)
-	}
+		return okResult
+	case ispec.MediaTypeImageManifest:
+		imgRes := CheckIntegrity(ctx, imageName, tag, manifest, imgStore)
+		listOfManifests[manifest.Digest] = imgRes
 
-	return results, nil
+		// if integrity ok then check subject if exists
+		if imgRes.Error == "" {
```
Comment (andaaron) [inline pkg/storage/scrub.go:187]: This code should be cleaner if we don't return results, just populate the map

**Comment by andaaron** [comment]
Why do we still have umoci in go.mod?

File: pkg/storage/scrub.go [hunk 3]
```diff
@@ -92,153 +91,234 @@ func CheckRepo(ctx context.Context, imageName string, imgStore storageTypes.Imag
...
+
+			var m ispec.Manifest
+
+			_ = json.Unmarshal(manifestContent, &m)
+
+			if m.Subject != nil {
+				imgRes = getManifestResult(ctx, *m.Subject, imgStore, imageName, tag, scrubbedManifests)
+				scrubbedManifests[manifest.Digest] = imgRes
+
+				return imgRes
```
Comment (andaaron) [inline pkg/storage/scrub.go:202]: Is this correct? We should be returning result for current image, not for the referred to image?

Comment (andaaron) [reply pkg/storage/scrub.go:202]: We are returning the result for the current image based on the imageName and tag parameters sent to getManifestResult at 199.

File: pkg/storage/scrub.go (hunk 1 above)
Comment (Andreea-Lupu) [reply pkg/storage/scrub.go]: Fixed. Now this function is called `scrubManifest` and it just adds the results to the map without returning them.

File: pkg/storage/scrub.go (hunk 2 above)
Comment (Andreea-Lupu) [reply pkg/storage/scrub.go:193]: Fixed.

**Review by andaaron — APPROVED** [review]

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**